src/hotspot/share/gc/g1/g1Policy.cpp

```diff
@@ -194,11 +194,14 @@ uint G1Policy::calculate_desired_eden_length_by_mmu() const {
 }
 
 void G1Policy::update_young_length_bounds() {
-  update_young_length_bounds(_analytics->predict_rs_length());
+  // We have no measure of the number of pending cards in the thread buffers,
+  // assume these are very few.
```

Comment:
`predict_pending_cards()` includes the cards that were thread buffers.  It is based on the cards that were processed, which is done after DCQS::concatenate_logs.

Reply:
Looking at the code:
`G1Policy::record_concurrent_refinement_stats` sets `_pending_cards_at_gc_start = dcqs.num_cards()`; `record_concurrent_refinement_stats` is called in `G1Policy::record_young_collection_start`; in `G1YoungCollector::collect`, `record_young_collection_start` is called _before_ `pre_evacuate_collection_set` which calls `G1BarrierSet::dirty_card_queue_set().concatenate_logs`.

I seem to be missing something, can you clarify?

Reply:
That's a bug.  I even found and fixed it in some earlier version of my concurrent refinement changes.  But it looks like I never extracted it out or even created a bug report for it, and there's no fix for it in my concurrent refinement PR.

Reply:
I'll file a bug for that problem.  I'd prefer this comment be removed, so it doesn't need to be tracked down for removal as part of fixing that bug.  Your call on that.